bulk_indexer.go

```diff
@@ -240,9 +255,74 @@ func (b *bulkIndexer) Flush(ctx context.Context) (BulkIndexerResponseStat, error
 	if err := jsoniter.NewDecoder(res.Body).Decode(&resp); err != nil {
 		return resp, fmt.Errorf("error decoding bulk response: %w", err)
 	}
+
+	for _, biri := range resp.FailedDocs {
+		if biri.Status == http.StatusTooManyRequests {
+			startlnIdx := biri.Position * 2
+			endlnIdx := startlnIdx + 2
+
+			if b.gzipw != nil {
+				gr, err := gzip.NewReader(bytes.NewReader(copyBuf))
+				if err != nil {
+					return resp, fmt.Errorf("failed to decompress request payload: %w", err)
+				}
+				defer gr.Close()
+
+				seen := 0
+
+				buf := make([]byte, 1024)
+				n, _ := gr.Read(buf[:1024])
+				buf = buf[:n]
+				for newlines := bytes.Count(buf, []byte{'\n'}); seen+newlines < startlnIdx; newlines = bytes.Count(buf, []byte{'\n'}) {
+					seen += newlines
+					n, _ := gr.Read(buf[:1024])
+					buf = buf[:n]
+				}
+
+				start := Indexnth(buf, startlnIdx-seen, '\n') + 1
+				end := Indexnth(buf, endlnIdx-seen, '\n') + 1
+
+				for end == 0 {
+					// Add more capacity (let append pick how much).
+					buf = append(buf, 0)[:len(buf)]
+
+					n, _ := gr.Read(buf[len(buf):cap(buf)])
+					buf = buf[:len(buf)+n]
+
+					end = Indexnth(buf, endlnIdx-seen, '\n') + 1
+				}
```

Comment:
The current test cases don't seem to cover the contents of the two `for` clauses. Would you mind adding a test or two where the bulk item is large enough to require more capacity?